python/dgl/heterograph.py

```diff
@@ -0,0 +1,1214 @@
+
+        Parameters
+        ----------
+        key : str or tuple
+            See above
+
+        Returns
+        -------
+        DGLBaseHeteroGraph
+            The induced subgraph.
+        """
+        pass
+
+
+class DGLHeteroGraph(DGLBaseHeteroGraph):
+    """Base heterogeneous graph class.
+
+    The graph stores nodes, edges and also their (type-specific) features.
+
+    Heterogeneous graphs are by default multigraphs.
+
+    Parameters
+    ----------
+    metagraph, node_types, edge_connections_by_type :
+        See DGLBaseHeteroGraph
+    node_frame : dict[str, FrameRef], optional
+        Node feature storage per type
+    edge_frame : dict[str, FrameRef], optional
+        Edge feature storage per type
+    readonly : bool, optional
+        Whether the graph structure is read-only (default: False)
+    """
+    # pylint: disable=unused-argument
+    def __init__(
+            self,
+            metagraph,
+            node_types,
+            edge_connections_by_type,
+            node_frame=None,
+            edge_frame=None,
+            readonly=False):
+        super(DGLHeteroGraph, self).__init__(metagraph, node_types, edge_connections_by_type)
+
+    # TODO: REVIEW
+    def add_nodes(self, num, node_type, data=None):
+        """Add multiple new nodes of the same node type
+
+        Parameters
+        ----------
+        num : int
+            Number of nodes to be added.
+        node_type : str
+            Type of the added nodes.  Must appear in the metagraph.
+        data : dict, optional
+            Feature data of the added nodes.
+
+        Examples
+        --------
+        The variable ``g`` is constructed from the example in
+        DGLBaseHeteroGraph.
+
+        >>> g['game'].number_of_nodes()
+        2
+        >>> g.add_nodes(3, 'game')  # add 3 new games
+        >>> g['game'].number_of_nodes()
+        5
+        """
+        pass
+
+    # TODO: REVIEW
+    def add_edge(self, u, v, utype, vtype, etype, data=None):
+        """Add an edge of ``etype`` between u of type ``utype`` and v of type
+        ``vtype``.
+
+        Parameters
+        ----------
+        u : int
+            The source node ID of type ``utype``.  Must exist in the graph.
+        v : int
+            The destination node ID of type ``vtype``.  Must exist in the
+            graph.
+        utype : str
+            The source node type name.  Must exist in the metagraph.
+        vtype : str
+            The destination node type name.  Must exist in the metagraph.
+        etype : str
+            The edge type name.  Must exist in the metagraph.
+        data : dict, optional
+            Feature data of the added edge.
+
+        Examples
+        --------
+        The variable ``g`` is constructed from the example in
+        DGLBaseHeteroGraph.
+
+        >>> g['user', 'game', 'plays'].number_of_edges()
+        4
+        >>> g.add_edge(2, 0, 'user', 'game', 'plays')
```

Comment:
How about `g['user', 'game', 'plays'].add_edges(2, 0)`?

Reply:
we probably don't allow mutating a subgraph.